File: Go/MinimumWindowSubsequence.go
//go:build ignore

package main

import "fmt"

func minWindow(s1, s2 string) string {
	m, n := len(s1), len(s2)
	i, j := 0, 0
	start, best := -1, m+1
	for i < m {
		if s1[i] == s2[j] {
			j++
			if j == n {
				right := i
				j--
				for j >= 0 {
					if s1[i] == s2[j] {
						j--
					}
					i--
				}
				i++ // left end of window
				if right-i+1 < best {
					best = right - i + 1
					start = i
				}
				j = 0
			}
		}
		i++
	}
	if start == -1 {
		return ""
	}
	return s1[start : start+best]
}

func main() {
	fmt.Println(minWindow("abcdebdde", "bde")) // bcde
	fmt.Println(minWindow("abc", "ac"))        // abc
	fmt.Println(minWindow("abcde", "xyz"))     // (empty)
}

File: Java/MinimumWindowSubsequence.java
public class MinimumWindowSubsequence {
    public String minWindow(String s1, String s2) {
        int m = s1.length(), n = s2.length();
        int i = 0, j = 0, start = -1, best = m + 1;
        while (i < m) {
            if (s1.charAt(i) == s2.charAt(j)) {
                if (++j == n) {
                    int right = i;
                    --j;
                    while (j >= 0) {
                        if (s1.charAt(i) == s2.charAt(j)) --j;
                        --i;
                    }
                    ++i;  // left end
                    if (right - i + 1 < best) {
                        best = right - i + 1;
                        start = i;
                    }
                    j = 0;
                }
            }
            ++i;
        }
        return start == -1 ? "" : s1.substring(start, start + best);
    }

    public static void main(String[] args) {
        MinimumWindowSubsequence s = new MinimumWindowSubsequence();
        System.out.println(s.minWindow("abcdebdde", "bde"));  // bcde
        System.out.println(s.minWindow("abc", "ac"));         // abc
        System.out.println(s.minWindow("abcde", "xyz"));      // (empty)
    }
}
